### PR TITLE
feat(python): add on_transform_error fault tolerance to StreamingDataset

### LINKED ISSUE #3704: Feature: StreamingDataset fault tolerance
### SDK

Python

### Description

Transformations can fail.  One very common reason is bad data (e.g. user survey where user skipped the question and we have a nan or null).  We should make it easy for users to skip these invalid rows during the loading process.

Open questions:

1. Can we do this?  Or do we need users to do the error handling in their transform function?
2. What do other libraries do for this?
3. Is this error handling on transform functions?  Or pre-filtering of invalid data (e.g. delete rows that have any nulls or nans)
4. How does this impact splits?  If one split suddenly has fewer rows due to bad data what will happen with things like elastic determinism?

### PR DESCRIPTION
**Pull request by Sravan1011**
Closes #3704

## Problem

Transforms can fail on bad data (e.g. nulls/NaNs from incomplete user surveys). Today any transform exception aborts iteration, and there is no way to skip invalid rows during loading.

## Solution

New `on_transform_error` parameter on `StreamingDataset`:

- `"raise"` (default, matches current behavior and the convention in tf.data / WebDataset / Ray Data)
- `"skip"` — drop the failing rows and continue
- `"warn"` — like skip, plus a logged warning per failing batch
- a WebDataset-style callable `handler(exc) -> bool`, so users can skip only expected error types

Key design points:

- **Row-granular skipping**: when a batch fails, the transform is re-run on single-row slices so only the rows that actually fail are dropped (avoids Ray-style whole-block loss). Skips are counted in a new `rows_skipped` property.
- **No crash on uneven skips**: the round-robin loop now ends the epoch at the last cycle where every split still has a row, instead of hitting `IndexError` when a split runs dry early.
- **Exact resumability under skips**: checkpoints are now position-based. `state_dict` gains `positions_consumed_per_split` (exact for owned splits), and a new `merge_state_dicts` static method combines per-rank states via elementwise max for elastic resume across topology changes. Old checkpoints without the new key still load. Positions equal sample counts when nothing is skipped, so existing behavior is unchanged.
- **Guardrail**: transforms returning the wrong number of rows now raise a clear `ValueError` instead of silently corrupting split accounting.

### Answers to the issue's open questions

- *Can we do this?* Yes — all transforms funnel through one guarded call in the Stage 2 pipeline.
- *What do other libraries do?* tf.data `ignore_errors()`, WebDataset `handler=`, Ray `max_errored_blocks`; MosaicML StreamingDataset offers nothing (skipping conflicts with its determinism model). This design follows the common conventions: raise by default, opt-in skipping, count/log drops.
- *Error handling or pre-filtering?* Both: the existing `filter=` remains the recommended tool for predictable bad data (splits are built post-filter, so all guarantees hold — now documented); `on_transform_error` covers failures not expressible as a predicate.
- *Impact on splits / elastic determinism?* Per-split sample sequences stay deterministic (skips are data-dependent, not topology-dependent). With unequal bad-row counts across splits the last few global steps of an epoch can differ across topologies (bounded by the skew), which is documented on the parameter. With equal counts per split, full determinism is preserved — covered by a test.

## Testing

15 new tests in `test_elastic_dataloader.py` covering: default raise, invalid values, uniform and uneven skips (including epoch-end truncation), warn logging, selective callable handlers, wrong-row-count guardrail, determinism across runs and across world sizes (1/2/3/4) with skips, exact mid-epoch resume with skips on the same topology, elastic resume via `merge_state_dicts` (ws=2 → ws=1), merge validation, and backward-compat loading of old checkpoints.

Note: relying on CI for the test run — my local machine OOMs during the final link of the native extension. The change itself is pure Python.

### CODE REVIEW

**Comment by Sravan1011** [comment]
@westonpace @chriscorrea can you review this

File: python/python/lancedb/streaming.py [hunk 1]
```diff
@@ -618,3 +785,59 @@ def load_state_dict(self, state: dict) -> None:
...
+        positions = state.get("positions_consumed_per_split")
+        if positions is None:
+            self._resume_positions = {}
+        else:
+            self._resume_positions = {
+                split: int(pos) for split, pos in enumerate(positions)
+            }
+
+    @staticmethod
+    def merge_state_dicts(states: list[dict]) -> dict:
```
Comment (westonpace) [inline python/python/lancedb/streaming.py:828]: I can see why this is needed but it is a fair bit of extra complexity.  I wonder if it would be possible to add an example here or just walk through something concrete.

For example, let's say I initially train on 8 ranks and then create a checkpoint and want to resume on 4 ranks.  Do I merge the state dicts from all 8 ranks and provide that identical merged dict to each of the 4 resume ranks?  Or do I need to somehow pairwise merge the 8 ranks so I end up with 4 ranks?  I think it's the former (since it's ok for the skip dict to have extra entries) but I want to verify and we should document it.

Also, what happens if I checkpoint 4 ranks and then expand to 8 ranks?  Again, I think I just merge all 4 and provide that identical state dict to each of the 8 ranks?

**Review by westonpace — APPROVED** [review]
This looks like a great addition.  Just one doc suggestion but we can merge provided gatekeeper is satisfied.  I have a few suggestions but they can be follow-up PRs.

* It might be nice to have a fill / placeholder option.  For example, an all-zero vector.  This could be a constant or a repair function.  It would have the nice property that it doesn't impact elastic determinism
* It might be good to have some kind of "max failures".  For example, if the transform function is just broken then it would scan the entire input and return 0 rows.  At some point we should probably just decide that things are broken and bail.

File: python/python/lancedb/streaming.py [hunk 2]
```diff
@@ -587,12 +739,27 @@ def state_dict(self) -> dict:
...
+        ``samples_consumed_per_split`` when ``on_transform_error`` skipped
+        rows, in which case entries are exact for the splits this instance
+        iterated and a lower bound (the sample count) for splits owned by
+        other ranks or workers.  Combine the state dicts from all ranks with
+        [merge_state_dicts][lancedb.streaming.StreamingDataset.merge_state_dicts]
+        to recover the exact value for every split before resuming on a
+        different topology.
         """
+        positions = [
+            self._resume_positions.get(split, self._resume_offset)
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/python/lancedb/streaming.py:782]: Hmm, I just recently added some tests for this case and I thought they were working.  This issue is unrelated to this PR, we can follow-up elsewhere.

File: python/python/lancedb/streaming.py [hunk 3]
```diff
@@ -407,15 +532,24 @@ def _ensure_cooked(i: int) -> None:
...
-                        # Stop when any split is exhausted (all exhaust
-                        # simultaneously: equal split sizes + round-robin).
-                        if any(local_consumed[i] >= split_sizes[i] for i in range(n)):
+                        # A cycle only runs if every split can still produce a
+                        # row.  Without skips all splits exhaust simultaneously
+                        # (equal split sizes + round-robin); when
+                        # on_transform_error drops rows a split can run dry
+                        # early, ending the epoch at the last complete cycle.
+                        exhausted = False
+                        for i in range(n):
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/python/lancedb/streaming.py:556]: It _would_ be nice if there was some cross-rank stop signal so that all ranks stopped early if one rank exhausted available rows.  However, I think that can be out of scope for this PR as long as the behavior is well documented.

File: python/python/lancedb/streaming.py [hunk 4]
```diff
@@ -522,6 +661,19 @@ def unscanned_rows(self) -> int:
...
+        """Number of rows dropped because their transform raised an exception.
+
+        Only ever non-zero when ``on_transform_error`` is set to ``"skip"``,
+        ``"warn"``, or a callable that returned ``True``.  Accumulates across
+        multiple iterations of the same dataset instance and is never reset
+        automatically.
+        """
+        if self._raw_batches_ref is not None:
+            return self._rows_skipped
+        return int(self._worker_stats[7])
```
[the remark this answers is not in the document]
Comment (westonpace) [reply python/python/lancedb/streaming.py:703]: Similar to the below comment, this doesn't seem to be unique to this PR.

**Comment by Sravan1011** [comment]
@westonpace I'll make the above changes then it will be a good merge right

**Comment by westonpace** [comment]
> I'll make the above changes then it will be a good merge right

Yeah, just ping me when you're ready.

**Comment by Sravan1011** [comment]
@westonpace  ready for another look, thanks for the review!

Changes since your review:

Documented the same-run rank imbalance: with on_transform_error != "raise", each rank stops based on its own splits only, so synchronous multi-rank training needs its own stop coordination (e.g. broadcasting on StopIteration). Also added a note at the exhaustion check in the code.
Added the worked example to merge_state_dicts — you were right, it's the former: merge all N previous-rank states into one dict and load that identical dict on every new rank, for both shrinking (8→4) and growing (4→8). No pairwise merging.
Fixed rows_skipped under-reporting on exhaustion (gatekeeper's finding): the shared-memory counter is now flushed when iteration ends, not just at cycle boundaries, so a split that dies before completing a cycle still reports its skips.
On gatekeeper's other two findings: the parent-process state_dict staleness with num_workers>0 looks pre-existing as you noted, and I've documented the cross-rank behavior rather than adding coordination, per your comment.

**Comment by Sravan1011** [comment]
Happy to take the fill/placeholder option and max_failures as follow-up PRs — want me to open issues for those?

**Comment by westonpace** [comment]
> Happy to take the fill/placeholder option and max_failures as follow-up PRs — want me to open issues for those?

Sure, that'd be great.  I'll merge this once CI is finished.

**Comment by Sravan1011** [comment]
> > Happy to take the fill/placeholder option and max_failures as follow-up PRs — want me to open issues for those?
> 
> Sure, that'd be great. I'll merge this once CI is finished.

could you do it again

**Comment by Sravan1011** [comment]
@westonpace done
